lib/iris/analysis/_area_weighted.py

```diff
@@ -1150,3 +459,266 @@ def regrid_callback(*args, **kwargs):
+    xy_rows = xy_rows.flatten()
+    xy_cols = (y_cols[:, np.newaxis] * x_src) + x_cols[np.newaxis, :]
+    xy_cols = xy_cols.flatten()
+
+    # Create a sparse matrix for efficient weight application.
+    combined_weights = csr_array(
+        (xy_weight, (xy_rows, xy_cols)), shape=(tgt_size, src_size)
+    )
+    return combined_weights
+
+
+def _standard_regrid_no_masks(data, weights, tgt_shape):
+    """
+    Regrid unmasked data to an unmasked result.
+
+    Assumes that the first two dimensions are the x-y grid.
+    """
+    # Reshape data to a form suitable for matrix multiplication.
+    extra_shape = data.shape[:-2]
+    data = data.reshape(-1, np.prod(data.shape[-2:]))
+
+    # Apply regridding weights.
+    # The order of matrix multiplication is chosen to be consistent
+    # with existing regridding code.
+    result = data @ weights.T
+
+    # Reshape result to a suitable form.
+    result = result.reshape(*(extra_shape + tgt_shape))
+    return result
+
+
+def _standard_regrid(data, weights, tgt_shape, mdtol):
+    """
+    Regrid data and handle masks.
+
+    Assumes that the first two dimensions are the x-y grid.
+    """
+    # This is set to keep consistent with legacy behaviour.
+    # This is likely to become switchable in the future, see:
+    # https://github.com/SciTools/iris/issues/5461
+    oob_invalid = True
+
+    data_shape = data.shape
+    if ma.is_masked(data):
+        unmasked = ~ma.getmaskarray(data)
+        # Calculate contribution from unmasked sources to each target point.
+        weight_sums = _standard_regrid_no_masks(unmasked, weights, tgt_shape)
+    else:
+        # If there are no masked points then all contributions will be
+        # from unmasked sources, so we can skip this calculation
+        weight_sums = np.ones(data_shape[:-2] + tgt_shape)
+    mdtol = max(mdtol, 1e-8)
+    tgt_mask = weight_sums > 1 - mdtol
+    # If out of bounds sources are treated the same as masked sources this
+    # will already have been calculated above, so we can skip this calculation.
+    if oob_invalid or not ma.is_masked(data):
+        # Calculate the proportion of each target cell which is covered by the
+        # source. For the sake of efficiency, this is calculated for a 2D slice
+        # which is then broadcast.
+        inbound_sums = _standard_regrid_no_masks(
+            np.ones(data_shape[-2:]), weights, tgt_shape
+        )
+        if oob_invalid:
+            # Legacy behaviour, if the full area of a target cell does not lie
+            # in bounds it will be masked.
+            oob_mask = inbound_sums > 1 - 1e-8
+        else:
+            oob_mask = inbound_sums > 1 - mdtol
+        # Broadcast the mask to the shape of the full array
+        oob_slice = ((np.newaxis,) * len(data.shape[:-2])) + np.s_[:, :]
+        tgt_mask = tgt_mask * oob_mask[oob_slice]
+
+    # Calculate normalisations.
+    normalisations = np.ones_like(weight_sums)
+    normalisations *= tgt_mask
```

Comment:
We believe this can be simplified using something like `normalisations = tgt_mask.copy`.